ada/ast.py

```diff
@@ -5213,8 +5213,24 @@ def is_derived_type(other_type=T.BaseTypeDecl.entity):
         other_can = Var(other_type.canonical_type)
         return Or(
             entity_can == other_can,
-            And(Not(entity_can.classwide_type.is_null),
-                entity_can.classwide_type == other_can.classwide_type),
+            Let(
+                lambda
+                entity_cw=entity_can.cast(ClasswideTypeDecl),
+                other_cw=other_can.cast(ClasswideTypeDecl):
+
+                Cond(
+                    Not(entity_cw.is_null) & Not(other_cw.is_null),
+                    entity_cw.typedecl.is_derived_type(other_cw.typedecl),
+
+                    Not(entity_cw.is_null),
```

Comment:
I'm not sure I understand this code, which means it needs comments :^)

Reply:
Nice comment! I re-read this code and thought of a way to improve it but it didn't work... Turns out there was a deeper problem which I didn't see originally! The new patch doesn't touch `is_derived_type` at all.